File: miscellaneous/matrix-dfs-traversal/20230114/matrix_dfs_stack.py
from typing import List, Tuple

DIRECTIONS = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def matrix_dfs(matrix: List[List[int]], start: Tuple[int]) -> List[int]:
    values: List[int] = []
    stack: List[Tuple[int]] = [start]
    seen: List[List[int]] = [[False for x in matrix[0]] for y in matrix]
    seen[start[0]][start[1]] = True

    while stack:
        row, col = stack.pop()
        values.append(matrix[row][col])

        for yOffset, xOffset in DIRECTIONS:
            newRow = row + yOffset
            newCol = col + xOffset

            if in_bounds(newRow, newCol, matrix) and not seen[newRow][newCol]:
                seen[newRow][newCol] = True
                stack.append((newRow, newCol))

    return values


def in_bounds(row: int, col: int, matrix: List[List[int]]):
    return 0 <= row < len(matrix) and 0 <= col < len(matrix[0])

File: miscellaneous/matrix-dfs-traversal/20230114/matrix-dfs-stack.php
<?php

declare(strict_types=1);

function matrixDfs(array &$matrix): array
{
    $values = [];
    $stack = [[0, 0]];
    $seen = array_fill(0, count($matrix), array_fill(0, count($matrix[0]), false));
    $seen[0][0] = true;
    $directions = [[-1, 0], [0, 1], [1, 0], [0, -1]];

    while ($stack) {
        [$row, $col] = array_pop($stack);
        $values[] = $matrix[$row][$col];

        foreach ($directions as [$offsetY, $offsetX]) {
            $newRow = $row + $offsetY;
            $newCol = $col + $offsetX;

            if (inBounds($newRow, $newCol, $matrix) && !$seen[$newRow][$newCol]) {
                $seen[$newRow][$newCol] = true;
                $stack[] = [$newRow, $newCol];
            }
        }
    }

    return $values;
}

function inBounds(int $row, int $col, &$matrix): bool
{
    return $row >= 0 && $row < count($matrix) && $col >= 0 && $col < count($matrix[0]);
}